File: routers/roadmap.py
"""Router for Teaching Roadmap endpoints (MCP tool handlers)."""

from fastapi import APIRouter, HTTPException
from typing import Optional
from pydantic import BaseModel
from services import db

router = APIRouter(prefix="/roadmap", tags=["roadmap"])

class StudiedUpdate(BaseModel):
    has_been_studied: bool

@router.get("")
async def get_teaching_roadmaps(course_name: Optional[str] = None):
    """Retrieve tracking info for all lessons in the Teaching Roadmap."""
    supabase = db.get_db()
    query = supabase.table("teaching_roadmap").select("*")
    
    if course_name:
        query = query.eq("course_name", course_name)
        
    try:
        res = query.order("date_of_lesson").execute()
        return res.data
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Database error: {exc}")


@router.post("/{lesson_id}/studied")
async def mark_lesson_studied(lesson_id: str, update: StudiedUpdate):
    """Mark a specific lesson as studied or not studied in the teaching roadmap."""
    supabase = db.get_db()
    try:
        res = supabase.table("teaching_roadmap").update({
            "has_been_studied": update.has_been_studied
        }).eq("id", lesson_id).execute()
        
        if not res.data:
            raise HTTPException(status_code=404, detail="Lesson not found.")
            
        return res.data[0]
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Database error: {exc}")

File: mcp-js/roadmap.js
import axios from "axios";

export const tools = [
    {
        name: "get_teaching_roadmaps",
        description: "Retrieve study tracking info for lessons in the Teaching Roadmap",
        inputSchema: {
            type: "object",
            properties: {
                course_name: { type: "string", description: "Optional. Filter by a specific course name" }
            },
            required: [],
        },
    },
    {
        name: "mark_lesson_studied",
        description: "Mark a specific lesson as studied or not studied in the teaching roadmap",
        inputSchema: {
            type: "object",
            properties: {
                lesson_id: { type: "string", description: "The UUID of the lesson" },
                has_been_studied: { type: "boolean", description: "Set to true if studied, false otherwise" }
            },
            required: ["lesson_id", "has_been_studied"],
        },
    }
];

export async function handleToolCall(name, args, API_BASE) {
    let res;
    switch (name) {
        case "get_teaching_roadmaps":
            let url = `${API_BASE}/roadmap`;
            if (args.course_name) {
                url += `?course_name=${encodeURIComponent(args.course_name)}`;
            }
            res = await axios.get(url);
            return res.data;
        case "mark_lesson_studied":
            res = await axios.post(`${API_BASE}/roadmap/${args.lesson_id}/studied`, {
                has_been_studied: args.has_been_studied
            });
            return res.data;
        default:
            return null;
    }
}
